x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1384,4 +1389,32 @@ protected LogicalPlan rule(Aggregate aggregate) {
             return plan;
         }
     }
+
+    /**
+     * Verify that a {@link LogicalPlan} can be executed.
+     *
+     * @param plan The logical plan to be verified
+     * @return a collection of verification failures; empty if and only if the plan is valid
+     *
+     * The following verifications on an optimized logical plan are supported:
+     * 1. EVAL with AUTO_BUCKET expression, its from and to should be foldable
+     */
+    LogicalPlan verify(LogicalPlan plan) throws VerificationException {
+        Collection<Failure> failures = new LinkedHashSet<>();
+        plan.forEachUp(p -> {
+            if (p instanceof Eval) {
+                p.forEachExpression(e -> {
+                    if (e.getVerifyOptimizedPlan()) {
+                        Expression.TypeResolution resolution = e.verifyOptimizedPlan();
+                        if (resolution.unresolved()) failures.add(fail(e, resolution.message()));
+                    }
```

Comment:
I don't think `e.getVerifyOptimizedPlan` (and its setter counterpart) are needed; this should do the same:

```suggestion
                    Expression.TypeResolution resolution = e.verifyOptimizedPlan();
                    if (resolution.unresolved()) failures.add(fail(e, resolution.message()));
```
(Per default, `e.verifyOptimizedPlan` already returns `TypeResolution.TYPE_RESOLVED`.)

Reply:
After reverting the changes to Expression.java in the last commit, this piece of code is outdated.